multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,65 +78,193 @@ func NewNodeReconciler(
 	klog.V(2).InfoS("Reconciling Node", "node", req.Name)
+	if !r.initialized {
+		if err := r.initialize(); err != nil {
+			return ctrl.Result{}, err
+		}
+		r.initialized = true
+	}
 	gw := &mcsv1alpha1.Gateway{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      req.Name,
 			Namespace: r.namespace,
 		},
 	}
-	// When the Node is annotated with 'multicluster.antrea.io/gateway=true' as a Gateway:
-	//   - Delete the Gateway if the Node is deleted
-	//   - Update the Gateway if Node's InternalIP or GatewayIP is updated
-	//   - Create a new Gateway if there is no existing Gateway
+
+	noActiveGateway := r.activeGateway == ""
+	isActiveGateway := r.activeGateway == req.Name
+	stillGatewayNode := false
+
 	node := &corev1.Node{}
 	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
-		if apierrors.IsNotFound(err) {
-			err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-			return ctrl.Result{}, client.IgnoreNotFound(err)
+		if !apierrors.IsNotFound(err) {
+			klog.ErrorS(err, "Failed to get Node", "node", req.Name)
+			return ctrl.Result{}, err
 		}
-		klog.ErrorS(err, "Failed to get Node", "node", req.Name)
-		return ctrl.Result{}, err
+	} else {
+		_, hasGWAnnotation := node.Annotations[common.GatewayAnnotation]
+		stillGatewayNode = hasGWAnnotation
 	}
 
-	_, isGW := node.Annotations[common.GatewayAnnotation]
-	var err error
-	gwNamespacedName := types.NamespacedName{
-		Name:      node.Name,
-		Namespace: r.namespace,
+	if stillGatewayNode {
+		r.gatewayCandidates[req.Name] = true
+	} else {
+		delete(r.gatewayCandidates, req.Name)
 	}
 
-	// TODO: cache might be stale. Need to revisit here and other reconcilers to
-	// check if we can improve this with 'Owns' or other methods.
-	var gwIP, internalIP string
-	if isGW {
-		if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
-			klog.ErrorS(err, "There is no valid Gateway IP for Node, will retry later when there is any new Node update", "node", node.Name)
-			return ctrl.Result{}, nil
+	var err error
+	var isValidGateway bool
+
+	if stillGatewayNode {
+		isValidGateway = true
+		gw.InternalIP, gw.GatewayIP, err = r.getGatawayNodeIP(node)
+		if err != nil {
+			klog.ErrorS(err, "There is no valid Gateway IP for Node", "node", node.Name)
+			isValidGateway = false
 		}
 	}
-	if err := r.Client.Get(ctx, gwNamespacedName, gw); err != nil {
-		if apierrors.IsNotFound(err) && isGW {
-			gw.GatewayIP = gwIP
-			gw.InternalIP = internalIP
-			if err := r.Client.Create(ctx, gw, &client.CreateOptions{}); err != nil {
+
+	if isActiveGateway {
+		if !stillGatewayNode || !isValidGateway || !isReadyNode(node) {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
 				return ctrl.Result{}, err
 			}
 			return ctrl.Result{}, nil
 		}
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		if err := r.updateActiveGateway(ctx, gw); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
```

Comment:
Maybe return before this line, as all other cases in this block return.

Reply:
Refined